src/connection.c

```diff
@@ -330,3 +328,24 @@ ConnectionGetOptionValue(PGconn *connection, char *optionKeyword)
 
 	return optionValue;
 }
+
+ShardId DoForAllShards(List* shardConnections, ShardAction action, void* arg)
```

Comment:
According to this logic everything should be placed in one function... Only in this case it should not be necessary to look into several places. 

DoForAllShards implements abstraction of iteration. So if programmer need to understand how shard placements are traversed, he/she should look only at this function. PgCopy*Transaction are operations performed with this shard placements. It has no relation with traversal rule.
Still do not think that elimination of generic iteration function is good idea.